### PR TITLE
Adding binary saving and loading to MLContext.Data

### LINKED ISSUE #1627: Binary Saving and Loading from MLContext
MLContext should have extensions for loading and saving binary files, like it currently has for text files. Specifically, add:

- `MLContext.Data.SaveAsBinary`, corresponding to `MLContext.Data.SaveAsText`
- `MLContext.Data.ReadFromBinaryFile`, corresponding to `MLContext.Data.ReadFromTextFile`

### PR DESCRIPTION
**Pull request by najeeb-kazmi**
Fixes #1627

### CODE REVIEW

File: src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    public static class BinaryLoaderSaverCatalog
+    {
+        /// <summary>
+        /// Read a data view from a binary file using <see cref="BinaryLoader"/>.
+        /// </summary>
+        /// <param name="catalog">The catalog.</param>
+        /// <param name="args">The arguments to binary reader.</param>
+        /// <param name="path"></param>
+        public static IDataView ReadFromBinaryFile(this DataOperations catalog,
+            BinaryLoader.Arguments args, string path)
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: >path [](start = 48, length = 4)

I think we should fix both this and `ReadFromText` to read from a stream.

File: src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+namespace Microsoft.ML
+{
+    public static class BinaryLoaderSaverCatalog
+    {
+        /// <summary>
+        /// Read a data view from a binary file using <see cref="BinaryLoader"/>.
+        /// </summary>
+        /// <param name="catalog">The catalog.</param>
+        /// <param name="args">The arguments to binary reader.</param>
+        /// <param name="path"></param>
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: >param [](start = 13, length = 5)

Add comment

File: src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs (hunk 1 above)
Comment (Zruty0) [inline src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: >args [](start = 35, length = 4)

Are there arguments to speak of? I'd suggest not exposing this parameter at all

File: src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            Contracts.CheckNonEmpty(path, nameof(path));
+
+            var env = catalog.GetEnvironment();
+
+            var reader = new BinaryLoader(env, args, path);
+            return reader;
+        }
+
+        /// <summary>
+        /// Save the data view as binary.
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: >as binary [](start = 31, length = 9)

into a binary stream

**Comment by Zruty0** [comment]
My approval is conditional on fixing all the open comments

File: src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+using Microsoft.ML.Runtime.Data.IO;
+
+namespace Microsoft.ML
+{
+    public static class BinaryLoaderSaverCatalog
+    {
+        /// <summary>
+        /// Read a data view from a binary file using <see cref="BinaryLoader"/>.
+        /// </summary>
+        /// <param name="catalog">The catalog.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: >The catalo [](start = 34, length = 10)

Maybe "The <see cref="DataOperations" /> catalog.

File: src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs (hunk 1 above)
Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: @Zruty0 I've changed this to read from a `Stream`. For `ReadFromTextFile`, I see that `TextLoader.Read` accepts `IMultiStreamSource` (of which `MultiFileSource` seems to be the only implementation that shows up in the codebase) and not `Stream`. Should I change `ReadFromTextFile` to read from a `MultiFileSource` instead?

Comment (Zruty0) [reply src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: For these extensions, use `Stream` (and not stream sources). Also, call them `ReadFromBla` where `Bla` is `Binary` or `Text`

---
In reply to: [235158400](https://github.com/dotnet/machinelearning/pull/1678#discussion_r235158400) [](ancestors = 235158400)

Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: There is no `TextLoader.Read` method that accepts a `Stream`. Do you suggest that we make `ReadFromText` accept a `FileStream`, then get the path to the file with `FileStream.Name`, and return 
```csharp 
reader.Read(new MultiFileSource(path)) 
```
?

Comment (Zruty0) [reply src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoaderSaverCatalog.cs]: Oh, right, I got it now. Ok, let's not touch `Text` then, and let's have `Binary` work from both streams and files.

**Pull request merged**